File: day14/day14.cpp
#include <algorithm>
#include <array>
#include <fstream>
#include <unordered_map>
#include <fmt/format.h>

using namespace std;

struct Rule
{
	int insert;
	size_t count;
	size_t other;
};

struct HashRule
{
	size_t operator()(const pair<int,int>& k) const
	{
		return (5381 * 33 + k.first) * 33 + k.second;
	}
};

using Ruleset = unordered_map<pair<int,int>,Rule,HashRule>;
using Letterset = unordered_map<int,size_t>;

static istream& parse_problem(istream& in, Ruleset &r, Letterset &letters)
{
	r.clear();
	letters.clear();

	enum { TEMPLATE, RULE } state = TEMPLATE;
	string line;
	string templ;
	while (getline(in, line))
	{
		if (state == TEMPLATE)
		{
			if (line.empty())
			{
				state = RULE;
			}
			else
			{
				templ = line;
			}
		}
		else if (!line.empty())
		{
			r[make_pair(line[0], line[1])].insert = line.back();
		}
		else
		{
			break;
		}
	}

	for (auto it = templ.begin(); it != templ.end(); ++it)
	{
		letters[(unsigned)*it]++;
		auto next = it + 1;
		if (next != templ.end())
		{
			r[make_pair(*it, *next)].count++;
		}
	}

	return in;
}

static size_t solve(Ruleset &r, Letterset &letters, size_t steps)
{
	while (steps-->0)
	{
		for (auto& [_, v] : r)
		{
			v.other = v.count;
			v.count = 0;
		}
		for (auto& [k, v]: r)
		{
			if (!v.other)
			{
				// empty
			}
			else if (!v.insert)
			{
				// doesn't transform
				v.count += v.other;
			}
			else
			{
				r[make_pair(k.first, v.insert)].count += v.other;
				r[make_pair(v.insert, k.second)].count += v.other;
				letters[v.insert] += v.other;
			}
		}
	}

	size_t max = 0;
	size_t min = SIZE_MAX;
	for (auto [_, x] : letters)
	{
		if (!x) { continue; }
		if (min > x) { min = x; }
		if (max < x) { max = x; }
	}
	return min < max ?  max - min : 0;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Letterset letterset;
	Ruleset ruleset;
	parse_problem(in, ruleset, letterset);
	in.close();

	fmt::print("Part1: {}\n", solve(ruleset, letterset, 10));
	fmt::print("Part2: {}\n", solve(ruleset, letterset, 30));

	return 0;
}

File: day14/day14.py
#!/usr/bin/env python3

def parse(txt):
    polymer = None
    rules = {}
    state = 0
    for line in txt.splitlines():
        if state == 0:
            if line == "":
                state = 1
            else:
                polymer = line
        else:
            pattern, insertion = line.split(" -> ")
            rules[pattern] = insertion

    return polymer, rules

def solve(initial, rules, steps):
    # populate the dictionary of the couples
    couples = {}
    for i in range(len(polymer)-1):
        couple = polymer[i:i+2]
        couples[couple] = couples.get(couple, 0) + 1

    # populate the dictionary of the letter counts
    letters = {}
    for l in polymer:
        letters[l] = letters.get(l, 0) + 1

    # perform the steps
    for i in range(steps):
        ncouples = {}
        for couple, count in couples.items():
            m = rules.get(couple)
            if m is None:
                # no rule applied
                ncouples[couple] = ncouples.get(couple, 0) + count
            else:
                a = couple[0] + m
                b = m + couple[1]
                ncouples[a] = ncouples.get(a, 0) + count
                ncouples[b] = ncouples.get(b, 0) + count
                letters[m] = letters.get(m, 0) + count

        couples = ncouples

    return max(letters.values()) - min(letters.values())

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        polymer, rules = parse(txt)

        print("Part1:", solve(polymer, rules, 10))
        print("Part2:", solve(polymer, rules, 40))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
